### PR TITLE
Feature/apvs 246/example worker process

### PR DESCRIPTION
**Pull request by tswann**
**Updated external web application**
- Generates a dummy reference number on application completed
- Creates a new task entry in mongodb
  - name = application-notification
  - status = PENDING
  - action = details required for sending notification to claimant
- Task queue with many consumers - this example for new application notifications only

**Example Node JS worker process**
- Runs on cron schedule to select PENDING notification tasks
- Sends mail for each tasks and then marks that task with a status of COMPLETED

**Debug SMTP server**
- Container running python debug SMTP server
- Service exposed on port 25, no auth required
- Spills message contents to console

### CODE REVIEW

File: alpha/notify-worker/app/process-notifications.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+var logger = require('./bunyan-logger')
+var client = require('./db-client')
+var mailer = require('./mailer')
+var Promise = require('bluebird')
+
+const EMAIL_CONTACT_METHOD = 'email'
+
+exports.run = function () {
```
Comment (stevenalexander) [inline alpha/notify-worker/app/process-notifications.js]: should have basic unit test covering that it calls to processNotification for all pending notifications

File: alpha/notify-worker/app/process-notifications.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+          processNotification(notification)
+        }))
+      }
+    })
+    .catch(function (error) {
+      logger.error(error)
+    })
+}
+
+var processNotification = function (notification) {
```
Comment (stevenalexander) [inline alpha/notify-worker/app/process-notifications.js]: Should we separated out, with process-notifications being the master/coordinator that gets the tasks from the DB and a separate module being the worker/actor which is called by the master to process them. This pattern would be easier to test and expand for multiple types of work:
http://blog.gigaspaces.com/the-master-worker-pattern/
http://www.michaelpollmeier.com/akka-work-pulling-pattern

File: alpha/notify-worker/app/config.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+
+module.exports = {
+  mail: {
+    host: 'smtp-server',
+    port: 25,
+    from: 'apvs@kainos.com'
+  }
+}
```
Comment (stevenalexander) [inline alpha/notify-worker/app/config.js:8]: Interesting approach, I could see us making config.js files in each of our significant applications, so all config values are in a single easy to find location. The values here could be set by environmental variable or scripted find/replace.
How do you see this working?

File: alpha/notify-worker/package.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
+{
+  "name": "notify-worker",
+  "version": "0.0.1",
+  "description": "",
+  "main": "app/app.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Comment (stevenalexander) [inline alpha/notify-worker/package.json]: we should be adding basic unit tests to all samples now

File: alpha/notify-worker/README.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# NOTIFY-WORKER
+
+This is an example async worker implemented as a node process.
+
+This example uses several node packages not used elsewhere in this proof of concept:
+
+* [cron](https://github.com/ncb000gt/node-cron): Allows for cron syntax scheduling of logic within the node application
+* [nodemailer](https://github.com/nodemailer/nodemailer): Email module for node js, used in this instance to send a simple plaintext message to the SMTP debug container
+
+### SMTP debug server
```
Comment (stevenalexander) [inline alpha/notify-worker/README.md]: should move some of this into the main README as it is included in the docker-compose

File: alpha/notify-worker/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,9 @@
+FROM node:6.5.0
+
+RUN mkdir -p /usr/src/app
+WORKDIR /usr/src/app
+
+COPY package.json /usr/src/app/
+COPY . /usr/src/app
+
+CMD npm install && node start.js
```
Comment (stevenalexander) [inline alpha/notify-worker/Dockerfile]: npm install should be moved to image build so dependencies are baked into image and not resolved at run time (only resolving deps at run time for development work)

File: alpha/external-web/app/routes/application-submitted.js [hunk 7]
```diff
@@ -1,6 +1,41 @@
 var router = require('../routes')
+var client = require('../services/eligibility-client')
+var logger = require('../services/bunyan-logger')
 
-router.get('/application-submitted', function (request, response, next) {
-  response.render('application-submitted')
+const tasksCollection = 'tasks'
+const claimantsCollection = 'claimants'
+
+router.get('/application-submitted/:claimant_id', function (request, response, next) {
```
Comment (stevenalexander) [inline alpha/external-web/app/routes/application-submitted.js]: creating notification should be done at last POST before application is completed.
A GET request should be Idempotent and repeatable.

File: alpha/notify-worker/app/config.js (hunk 3 above)
Comment (tswann) [reply alpha/notify-worker/app/config.js:8]: Environment variables was what I had in mind - I've set these before from TeamCity as part of the CI process for particular environments, assuming Jenkins can do something similar - perhaps one to check in the coming sprint.

File: alpha/notify-worker/Dockerfile (hunk 6 above)
Comment (tswann) [reply alpha/notify-worker/Dockerfile]: Do you mean replacing with a

```
RUN npm install
```

prior to the command/entry point?

**Comment by tswann** [comment]
Hi @stevenalexander - review comments applied

**Pull request merged**